### ISSUE TITLE
Customized ogx360 hardware with Pro Micro 3.3V

### ISSUE THREAD

**Issue by phantom-8** [Ryzee119/ogx360#41]
I've built a simplified version of ogx360 with a Pro Micro 3.3V/8MHz board and a Mini USB Host Shield 2.0.  It only supports 1 Xbox 360 controller.  Everything seems to work fine, except it missed a button release event once a while.  I suspect the culprit is the slower 8MHz of the Pro Micro.  Do you have any idea on how to tackle the problem?

![IMG_20201118_140232](https://user-images.githubusercontent.com/8468733/99492031-cff2b880-29a7-11eb-8f0c-f7cd9a080012.jpg)

**Comment by Ryzee119** [comment]
Very nice!
It could certainly be an issue. As its limited to one controller you can try these:

Make this loop to 1 instead of 4 to avoid some overhead
https://github.com/Ryzee119/ogx360/blob/27942f1965dc4b0a7e773592d668ba88c2513541/Firmware/ogx360_32u4/main.cpp#L223

Have you adjusted `F_CPU` to match your new clock? Not sure how your compiling, but its a define setup in your project.

You could try bumping up the SPI speed. I think its set here (and look for other instances for 12000000 in that file. The default was originally 26000000.

https://github.com/Ryzee119/ogx360/blob/27942f1965dc4b0a7e773592d668ba88c2513541/Firmware/ogx360_32u4/lib/USB%20Host%20Shield/usbhost.h#L62

**Comment by phantom-8** [comment]
Thank you so much for the prompt reply.  I've already removed unnecessary code in main.cpp.  At the moment, all I2C code are gone; the main "for loop" iterates only once; master's slave handling code are gone.  The F_CPU is already set to 8000000UL.  Unfortunately, the problem still exist.  

I'm gonna try to play around with the SPI setClockSpeed and see if it helps.  Thanks again for your kind assistance.

**Comment by phantom-8** [comment]
Just tested the SPI Clock.  In the ushhost.h, the ATMega32u4 doesn't use setClockSpeed.  Instead, the clock speed is being set via SPCR & SPSR registers.
```
78                 SPCR = 0x50; //SPI Enable and MASTER,
79                 SPSR = 0x00; // 0x01 = SPI2X Ryzee119, changed from 0x01 to 0x00 to disable SPI2X. Speed is fclk/4 now.
```
The SPI Clock is currently being set to 4MHz for Pro Micro 16MHz version, and 2MHz for Pro Micro 8MHz version.  I've bumped the clock speed to 4MH by setting SPSR=0x01.  The 4MHz SPI Clock is the maximum that can be set on the Pro Micro 8MHz version.  Unfortunately, this didn't fix the problem.

Can I ask how the MAX3421 handles controller button events?  Does it only send button change events via SPI (that is, only send 1 event for button down and another event for button up)?

**Comment by Ryzee119** [comment]
ah yes ofcourse. been a while since Ive looked deep into this code.

MAX3421 generates an interrupt on all new events on the usb device. ogx360 polls this interrupt register to detect changes and applies them accordingly.

Most controllers will only send a new packet if a change has been detected. For example if you're moving the analog stick around do the missed button presses happen?

Im wondering if this section is happening *too quickly* and is clearing previous buttons before it can send it to the og xbox. ogxbox is capped as 4ms per interrupt packet. xbox 360 can do 1ms.

https://github.com/Ryzee119/ogx360/blob/27942f1965dc4b0a7e773592d668ba88c2513541/Firmware/ogx360_32u4/main.cpp#L230-L238

Although counter intuitive, add a 4ms delay on each loop.

The 4ms thing is here. You could try changing the `>=4` to 0 or 1. Although I have had issues when im spamming the og too fast much. :)
https://github.com/Ryzee119/ogx360/blob/27942f1965dc4b0a7e773592d668ba88c2513541/Firmware/ogx360_32u4/main.cpp#L688

**Comment by phantom-8** [comment]
> MAX3421 generates an interrupt on all new events on the usb device. ogx360 polls this interrupt register to detect changes and applies them accordingly.
> 
Thanks for the explanation.  

> Most controllers will only send a new packet if a change has been detected. For example if you're moving the analog stick around do the missed button presses happen?
> 
Yes, if I moved the analog stick, the missed button press/release will be updated.

> Im wondering if this section is happening _too quickly_ and is clearing previous buttons before it can send it to the og xbox. ogxbox is capped as 4ms per interrupt packet. xbox 360 can do 1ms.
> 
> Although counter intuitive, add a 4ms delay on each loop.
> 
> The 4ms thing is here. You could try changing the `>=4` to 0 or 1. Although I have had issues when im spamming the og too fast. :)
> 
> 
It seems to me your code is fine.  I don't think the problem is related to your button handling or the timing on sending HID report to OG.  My theory is that the 8MHz Pro Micro is only running at half speed, it's not fast enough to poll the interrupt register to capture all button change events.  I think the following are the possible ways to solve the problem:

1) Use a Pro Micro version with 3.3V/16MHz which are being sold on ebay & Aliexpress.  The ATMega32u4 will be running out of spec, but most people reports are positive.

2) The current implementation uses polling to process the button events.  This will be problematic for slower MPU.  A better way would be interrupt driven.  I think you mentioned to migrate to UHS3.0 in the future.

3) UsbHost.Task() is not being called fast enough and result in lost events.  Currently, it's being called once at the beginning of the main loop.  We can try to add 1 or more calls to UsbHost.Task() in the main loop and spread them evenly. 

I will try approach 3 first and see if it helps.  If not, I think I may go with approach 1.

**Comment by phantom-8** [comment]
Approach 3 didn't work as the problem is within UsbHost.Task().  The Poll() function makes calls to checkControllerPresence(), checkControllerBattery(), setLedRaw(), chatPadKeepAlive1() & chatPadKeepAlive2() every 2.5 seconds.  These functions will send commands to MAX3421 via SPI bus.  As long as I comment out all these functions, the system will run reliably and no more message lost.  If I leave any 1 of these functions intact, it will also result in message lost.  These functions create more traffic on the SPI bus and the 8MHz Pro Micro seems insufficient to handle it.  I guess my only option now is to use a Pro Micro version with 3.3V/16MHz.

https://github.com/Ryzee119/ogx360/blob/27942f1965dc4b0a7e773592d668ba88c2513541/Firmware/ogx360_32u4//lib/USB%20Host%20Shield/XBOXRECV.cpp#L280-L315

**Comment by jimnarey** [comment]
Hi Phantom8 - is there any chance you could share your simplified code? I was about to dive in and do something similar myself (so far using a Leonardo with a full sized shield).

Thanks.

**Comment by Ryzee119** [comment]
https://github.com/Ryzee119/ogx360_t4
I knocked this up recently.
Which uses a single teensy4.1 that can do dual usb. 
Its a huge jump in power and flash memory so opens up lots of possibilities for future work.

**Comment by Ryzee119** [comment]
![126860412_2475733002727915_8715832948255045769_n](https://user-images.githubusercontent.com/21236406/100154560-b268a600-2ef5-11eb-9a0d-244ba176b211.jpg)

**Comment by jimnarey** [comment]
And based in VS Code! Thank you so much for pointing me to this.

**Comment by phantom-8** [comment]
I've built another version based on Pro Micro 5V/16MHz.  I used 5V to 3.3V logic convertors for SS, MOSI, SCLK & D9.  Basically, the circuit is the same as the one on your schematics.  Unfortunately, this version also encounters button lost events, exactly like the 3.3V/8MHz version.  I tried using both your Github compiled firmware and also my own built.  They've the same problem.  I'm really puzzled on what's going on.  

@Ryzee119, is it possible for you to test whether your version has button lost events?  The way that I tested was using the Controller test tool from your Video.  I just keep pressing/releasing the same button and watch the controller screen for missing press/release.  In average, I would lose a button event every 100 button presses.

@jimnarey, do you still need the trimmed down version from me?  Let me know if you do and I'll cleanup the source code and upload to Github as a fork.

**Comment by Ryzee119** [comment]
Quickly have a shot if this
[ogx360_32u4_master.zip](https://github.com/Ryzee119/ogx360/files/5594796/ogx360_32u4_master.zip)

**Comment by Ryzee119** [comment]
Does it happen on D-Pad and ABXY actually?

**Comment by Ryzee119** [comment]
Desregard previous, try this one
[ogx360_32u4_master.zip](https://github.com/Ryzee119/ogx360/files/5594835/ogx360_32u4_master.zip)

**Comment by phantom-8** [comment]
> Desregard previous, try this one
> [ogx360_32u4_master.zip](https://github.com/Ryzee119/ogx360/files/5594835/ogx360_32u4_master.zip)

Do u want me to try this version?

I forgot to mention that I was using a Xbox 360 Wireless controller with Chatpad.  I believe it will be easier to reproduce with Chatpad connected.

**Comment by Ryzee119** [comment]
ooof yea you forgot to mention that :P Is this for steel battalion?

**Comment by phantom-8** [comment]
> > Desregard previous, try this one
> > [ogx360_32u4_master.zip](https://github.com/Ryzee119/ogx360/files/5594835/ogx360_32u4_master.zip)
> 
> Do u want me to try this version?
> 
> I forgot to mention that I was using a Xbox 360 Wireless controller with Chatpad. I believe it will be easier to reproduce with Chatpad connected.



> ooof yea you forgot to mention that :P Is this for steel battalion?

Yeah, it's for Steel Battalion.  BY THE WAY, it also happened with Chatpad when I was using Pro Micro 3.3V/8MHz.  I've test without Chatpad on 5V/16MHz version yet.

**Comment by Ryzee119** [comment]
All good I can replicate now. Ill see if I can fix

**Comment by phantom-8** [comment]
> All good I can replicate now. Ill see if I can fix

Great!  I thought I was doing something wrong.  :)

**Comment by Ryzee119** [comment]
Im struggling to fix this. Its almost like the chatpad packets are overwriting the button packets. i dont know know. This could be an inherent issue with Usb Host Shield v2 :(

**Comment by phantom-8** [comment]
When I was testing the old Pro Micro 3.3V/8MHz, I also experienced button event lost even without chatpad connected.  However, the problem occurrence is much less frequent without chatpad.  I will do some more testing on the 5V/16MHz version to confirm whether the problem also existed without chatpad.

**Comment by phantom-8** [comment]
I've just tested the 5V/16MHz version with a Xbox 360 Wireless Controller without chatpad connected.  The button event lost also occurred, but it's very difficult to reproduce and took me 7 minutes to get 1 lost event.  To make it easier to reproduce, you can modify the timer in the Poll() function from 2500ms to let's say 200ms.  Then, the problem will occur much more frequent.  From my experience, as long as you send any command to the controller over SPI bus in the Poll() function, there is a possibility that a button event may be lost.

https://github.com/Ryzee119/ogx360/blob/27942f1965dc4b0a7e773592d668ba88c2513541/Firmware/ogx360_32u4/lib/USB%20Host%20Shield/XBOXRECV.cpp#L280-L286

**Comment by phantom-8** [comment]
I'm thinking whether it's possible to manually request a HID report from the controller.  If it's possible, maybe a timer can be added to regularly request additional HID report in order to bypass the missing button event issue.

**Comment by Ryzee119** [comment]
Yea it mighy be possible to request button state. Not sure of the command to do that yet

**Comment by phantom-8** [comment]
I searched the net, but couldn't find the manual request button state command.  If you happen to find it, please post here.  I can try to add the code and test if it works.

**Comment by jimnarey** [comment]
@phantom-8 - I've pared back the code myself, but thank you! You can see the repo here if you're interested: https://github.com/jimnarey/SimpleXboxControllerAdapter

**Comment by phantom-8** [comment]
@jimnarey, excellent work!  You've provided a very detailed description on how to build a simple version with off-the-shelf parts.  My interest is to use a Xbox 360 Wireless controller with Chatpad to emulate Steel Battalion controller.  I've the original full size controller, but will like to play SB LOC with my daughter.  That's why I'll like to fix the message lost problem.  I'm gonna try to use a STM32F4 to test whether there will be any message lost for a faster version of the the MCU.  I suspect the problem is caused by the slow Pro Micro MCU.

**Comment by jimnarey** [comment]
Thank you! I'm now making a bit of progress implementing PS3 support but am having some problems compiling the project with PS4 support. There's something a bit strange going on. How are you compiling yours?

**Comment by phantom-8** [comment]
I'm compiling under MacOS using CMake & avr-gcc.  I followed instructions here.
https://github.com/Ryzee119/ogx360/pull/32

**Comment by jimnarey** [comment]
Thank you. I've had a look at the issue/instructions but I can't see from them how to compile locally. I suspect, though, my problem is something a bit more fundamental as I was able to reproduce it in Microchip Studio. I've opened another issue. It's quite odd, I can add in the PS3 controller libraries perfectly well without the linker complaining.

If you could let me know what you're running on MacOS and you don't mind, I'll add it to my guide. At the moment it's relying on VSCode to run the CMake tasks.

**Comment by phantom-8** [comment]
The is a shell script auto-build file under
ogx360/.github/workflows/ci_build.sh

Just copy the script to your main ogx360 directory and execute the script.  It should automically build the firmware .

**Comment by jimnarey** [comment]
How are you getting on with the lost button presses? 

With a Leonardo and USB Host Shield (so 5V, 16Mhz) using wired controllers only (Xbox One, 360, PS3, PS4), I haven't noticed any missing button presses yet but testing has basically involved a bit of Halo to check button mapping/for latency and a bit of Burnout 3 to check R1/R2 pressure is being handled properly. I'm happy to do further testing if that would be helpful.

I'm about to build one using a Pro Micro at 5V/16Mhz, 3.3V regulator, level shifter and mini shield. I'll see how that works.

I've been looking back through this discussion and noted the bit about the Poll() function in XBOXRECV. The equivalent in XBOXUSB is much simpler:

uint8_t XBOXUSB::Poll() {
        if(!bPollEnable)
                return 0;
        uint16_t BUFFER_SIZE = EP_MAXPKTSIZE;
        pUsb->inTransfer(bAddress, epInfo[ XBOX_INPUT_PIPE ].epAddr, &BUFFER_SIZE, readBuf); // input on endpoint 1
        readReport();
#ifdef PRINTREPORT
        printReport(); // Uncomment "#define PRINTREPORT" to print the report send by the Xbox 360 Controller
#endif
        return 0;
}

Could you meet your overall aim with a chatpad and wired controller?

**Comment by phantom-8** [comment]
I've done some testing with ESP32 which is a lot faster than Pro Micro.  Unfortunately, the lost event also occurred in this version.  Here is what I've found so far:

- My previous statement of sending any commands over SPI bus will cause button event lost is not totally correct.  It's not the SPI send commands that cause button lost.  It's the reply resulted from those commands that cause button event message lost/overwritten.
- In the Poll() function, the following commands will result in replies from Controller.  Only replies with 2 or more consecutive messages will overwrite button events.  Command with a single reply message works fine.
  * checkControllerPresence(): Reply with 2 messages (1 short & 1 long).
  * checkControllerBattery(): Reply with 1 long message
  * setLedRaw(): No reply message
  * chatPadKeepAlive1(): Reply with 2 or more long messages
  * chatPadKeepAlive2(): Reply with 2 or more long messages
- The commands checkControllerPresence(), chatPadKeepAlive1() & chatPadKeepAlive2() will result in 2 or more consecutive reply messages and will cause button event lost/overwritten.  The possibilities of event lost depends on number of messages on the reply.  The more the reply messages, the higher the change of losing the button events.  Therefore, checkControllerPresence() has less chance of overwritting button events comparing to chatPadKeepAliveN().
- In order to reproduce the problem easier, I've change the checkStatusTimer from 2500ms to 10ms; as this will create a lot more reply messages.
- On ESP32, I've changed the SPI speed from 4MHz to 26MHz which is the maximum supported by MAX3421E.  Problem still exist.
- I've also tried to connect 2 wireless controllers.  On controller 1, Pro Micro keeps calling chatPadKeepAliveN() in every 10ms.  On controller 2, Pro Micro won't send any SPI commands and I continuously hit the same button on Controller 2 to test for event lost.  Under this configuration, there are no message lost/overwritten.  Thus, the problem has nothing to do with the bandwidth of the SPI bus, since the same bus still has a lot of reply messages from controller 1.  The difference is that they're on 2 different endpoints.

@jimnarey, the reason the Poll() function in XBOXUSB is much simpler is because it doesn't support Chatpad.  In fact, I'm in the process of adding Chatpad support to the XBOXUSB (Wired Controller).  From my test with 2 controllers above, it's very likely the button event lost/overwritten problem won't occur on XBOXUSB because the Chatpad is using a different endpoint.  As for the Wireless Controller, the same endpoint is being shared by controller button events & chatpad events and this is when messages got lost.

if you wanna simulate the problem, you can simply change the checkStatusTimer from 2500ms to 10ms in the Poll() function and continuously hit the same button.  Then, you should be able to reproduce it.

**Comment by phantom-8** [comment]
> I'm about to build one using a Pro Micro at 5V/16Mhz, 3.3V regulator, level shifter and mini shield. I'll see how that works.

I'm still waiting for my Pro Micro 3.3V/16MHz to arrive.  If it works, we don't need level shifters and the build is much easier.  BY THE WAY, I think 3.3V/8MHz works fine as the message lost problem is not caused by the slower MCU.  However, it's nicer to have a faster MCU. :P

**Comment by Ryzee119** [comment]
@phantom-8 i think is onto it. 
Ive been trying and I simply cant fix it. I dont know if its a limit of UHS2 or the MAX3421, or we're missing something in our understanding of the protocol all together.
(Or it actually happens on the real xbox 360)

**Comment by phantom-8** [comment]
@Ryzee119, I noticed that on ogx360_t4, you're emulating Steel Battalion controller with mouse & keyboard.  Do you have any plans to add support to Xbox 360 Wireless Controller with Chatpad.  I'm interested to know if the message lost problem still exist on the new platform.

**Comment by Ryzee119** [comment]
Yes i plan to. Im also interested 😊

**Comment by phantom-8** [comment]
@Ryzee119, I noticed there are 2 versions of Xbox 360 Wired Controller, 1.10 & 1.14.  The one I have is 1.14.  The endpoint numbering seems to be different.  In your XBOXUSB.cpp, you're using endpoint address 0x02 for the controller output pipe.  Is your controller version 1.10?  I need to change the output pipe endpoint address to 0x01 for mine to work.

https://github.com/Ryzee119/ogx360/blob/27942f1965dc4b0a7e773592d668ba88c2513541/Firmware/ogx360_32u4/lib/USB%20Host%20Shield/XBOXUSB.cpp#L146-L157

I've gotten the chatpad to respond and it's a pain in the ass to get the timing right.  Will test for missing button events later.

**Comment by Ryzee119** [comment]
Yea must be different versions. Mine is definitely pipe 0x02.
Is yours both on pipe 0x01?

The correct way is to parse the configuration descriptor to pull out the endpoint addresses.
It gets the device descriptor here. There another function called `getConfDescr` which I think might work.

Unfortunately these shortcuts exist to save space on the tiny microcontroller. I didnt know there was another version until now

**Comment by phantom-8** [comment]
Yeah, both my Controller Input & Output pipes are 0x01.  The Chatpat endpoints are also different for the 2 versions.  Need to add special handling.

**Comment by Ryzee119** [comment]
Ive tried get UHS v3 running on the ogx360 which doesnt have any of these shortcuts. Ive got it going but constantly runs out of memory. I have a feeling the limited 32u4 just cant cut it :(

**Comment by phantom-8** [comment]
> Ive tried get UHS v3 running on the ogx360 which doesnt have any of these shortcuts. Ive got it going but constantly runs out of memory. I have a feeling the limited 32u4 just cant cut it :(

Is it possible to trim down the code just enough to test the Xbox 360 Wireless controller with Chatpad to see if there are any lost button events on UHS v3?

I've just finished testing for button event lost with XBOXUSB.  There're aboslutely no button event lost, no matter how hard I jam the bus with Chatpad traffic.  This is expected as the chatpad & the controller are on different endpoints.  I still have some work to do to finish up the chatpad & Steel Battalion coding on the XBOXUSB.  When I'm done, I'll create a pull request.

**Comment by Ryzee119** [comment]
I have found something interesting!

Try changing this line
https://github.com/Ryzee119/ogx360/blob/27942f1965dc4b0a7e773592d668ba88c2513541/Firmware/ogx360_32u4/lib/USB%20Host%20Shield/XBOXRECV.cpp#L377

to this

`if(readBuf[0] == 0x00 && (readBuf[1] == 0x01 || readBuf[1] == 0x03))`

When mashing the chatpad and buttons at the same time, theres is new packet with a 0x03 ID. I cant see this documented anywhere online.

Seems to have fixed my problems!

**Comment by Ryzee119** [comment]
I got the chatpad going of my ogx360_t4 variant and it was doing the same thing. Saw these unusual packets and looked into it.
In hindsight this rings a bell.. damn.

I havent tested this on traditional ogx360 yet. so please let me know

**Comment by phantom-8** [comment]
Nice catch!  Microsoft combined the button & chatpad packets into a single one!  In the 1st test, I continuously performed controller button bashing with chatPadKeepAliveN() running every 10ms.  It created lots of 0x03 packets and this is in fact the cause of missing button presses with chatpad ’cos we didn’t check for the 0x03.  lol

Unfortunately, there is a catch.  As I mentioned earlier, besides chatPadKeepAliveN(), checkControllerPresence() will also result in 2 consecutive reply messages.  My 2nd test is to perform controller button bashing with checkControllerPresence() running every 10ms.  This time a button lost event really occurred.  Here is the log:

```
[00:01:17.049] Controller 0: 08 80 
[00:01:17.051] Controller 0: 00 0F 00 F0 00 CC FD C3 97 D0 74 4C 78 D0 00 20 1B A3 20 1D 30 03 40 01 50 01 FF FF FF
[00:01:17.060] Controller 0: 08 80 
[00:01:17.062] Controller 0: 00 0F 00 F0 00 CC FD C3 97 D0 74 4C 78 D0 00 20 1B A3 20 1D 30 03 40 01 50 01 FF FF FF
[00:01:17.066] Controller 0: 00 01 00 F0 00 13 00 10 00 00 E8 00 AD F6 66 10 7C FA 00 00 00 00 00 00 00 00 00 00 00
[00:01:17.071] Controller 0: 08 80 
[00:01:17.073] Controller 0: 00 0F 00 F0 00 CC FD C3 97 D0 74 4C 78 D0 00 20 1B A3 20 1D 30 03 40 01 50 01 FF FF FF
[00:01:17.078] Controller 0: 00 00 00 F0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
[00:01:17.083] Controller 0: 08 80 
[00:01:17.085] Controller 0: 00 0F 00 F0 00 CC FD C3 97 D0 74 4C 78 D0 00 20 1B A3 20 1D 30 03 40 01 50 01 FF FF FF
[00:01:17.094] Controller 0: 08 80 
[00:01:17.096] Controller 0: 00 0F 00 F0 00 CC FD C3 97 D0 74 4C 78 D0 00 20 1B A3 20 1D 30 03 40 01 50 01 FF FF FF
```

```
[00:01:17.315] Controller 0: 08 80
[00:01:17.317] Controller 0: 00 0F 00 F0 00 CC FD C3 97 D0 74 4C 78 D0 00 20 1B A3 20 1D 30 03 40 01 50 01 FF FF FF
[00:01:17.326] Controller 0: 08 80
[00:01:17.328] Controller 0: 00 0F 00 F0 00 CC FD C3 97 D0 74 4C 78 D0 00 20 1B A3 20 1D 30 03 40 01 50 01 FF FF FF
[00:01:17.337] Controller 0: 00 00 00 F0 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
[00:01:17.341] Controller 0: 08 80
[00:01:17.348] Controller 0: 08 80
[00:01:17.350] Controller 0: 00 0F 00 F0 00 CC FD C3 97 D0 74 4C 78 D0 00 20 1B A3 20 1D 30 03 40 01 50 01 FF FF FF
[00:01:17.359] Controller 0: 08 80
[00:01:17.361] Controller 0: 00 0F 00 F0 00 CC FD C3 97 D0 74 4C 78 D0 00 20 1B A3 20 1D 30 03 40 01 50 01 FF FF FF
```
At time 00:01:17.066, button A has been pressed and detected.  At 00:01:17:337, you’ll see the message “00 00 00 0F 00…”.  There should be a button release message before this one and it’s missing.  However, the occurrence of this is extremely rare and it’s very difficult to reproduce when calling checkControllerPresence() every 2.5 seconds.  I recalled that it took me 7 minutes to got 1 lost event last time.  Thus, I wouldn’t worry about fixing this one.

Anyway, I’m happy with the Chatpad fix and thank you so much for the hard work!  :)

If you're interested, you can download the above log file for the 2nd test here
https://drive.google.com/file/d/13SH5V3YFz0XbzkvY0QnGK6sydX_0xA3i/view?usp=sharing.

**Comment by Ryzee119** [comment]
I made alot of improvements/changed a few things during this investigation in ogx360 code. It may help these last few lost packets. Ill clean it up and push it as soon a I can

**Comment by phantom-8** [comment]
Great!  Once you push the changes, I can help to retest.  Thanks again!

**Comment by Ryzee119** [comment]
https://github.com/Ryzee119/ogx360/tree/xboxrecv/Firmware/ogx360_32u4
Try this branch. Needs to be burn tested will be interesting to see if this helps with the rare button press you are missing now

**Comment by phantom-8** [comment]
In source file XBOXRECV.cpp, you've added the handling of 0x03 packet for button events, but you haven't handled 0x03 on the chatpad.  The 0x03 is a merged packet for controller & chatpad events.  I would suggest change line #373 to this in order to avoid missing chatpad events.

```
    } 
    if(readBuf[0] == 0x00 && (readBuf[1] == 0x02 || readBuf[1] == 0x03)){
```

I'll preform testing in a moment.

**Comment by Ryzee119** [comment]
the `0x03` packet I think only contains the chatpad led status, not chatpad button events from what I have seen, but I could be wrong though

**Comment by phantom-8** [comment]
It's most likely the Controller firmware merged button & chatpad events due to traffic concern.  If 0x03 can contain chatpad LED events, it's very likely it may also contain chatpad buttons.

**Comment by Ryzee119** [comment]
yep looks like youre right :) Ill push a new change shortly

**Comment by Ryzee119** [comment]
Alright, ive push those changes. Disclaimer havent tested them myself

**Comment by phantom-8** [comment]
Will test the latest update now.

**Comment by Ryzee119** [comment]
sorry, just found another small bug which I have fixed related to chatpad returning incorrect buttons if you mashed multiple buttons quick enough  👍 
Pushed that too.

Nice this seems to be working alot better now

**Comment by phantom-8** [comment]
No problem.  Will fetch the update.  :)

**Comment by phantom-8** [comment]
Time for test report, here are the findings:

1) When pressing the Guide button on the controller to connect to the wireless receiver, they sync up perfectly every time.  However, the controller number LED indicator sometimes keeps flashing (4 ring LEDs keep flashing) without settling to show the controller number.  To reproduce, press Guide to connect.  Then, press and hold Guide for a few seconds to disconnect.  I’ve repeatedly tested 10 times, 4 out of 10 will just keep flashing.  Other 6 times successfully settled to show controller number LED.  BY THE WAY, I’ve re-tested the original firmware before these code changes, this problem wasn’t there.

2) With the stock firmware, that is checkStatusTimer at 2500ms, I’m unable to reproduce button lost.  PERFECT!!!

3) Modified firmware with checkStatusTimer at 10ms and only calls to checkPadKeepAliveN(), no lost button event detected!

4) Modified firmware with checkStatusTimer at 10ms and only calls to checkControllerPresence(), lost button event occurred with button bashing.  With the attached log file, at time 00:00:15:183, controller button A press was deteced.  But no button release event was deteced afterwards.

The checkControllerPresece() still results in lost message under extreme condition with 10ms calling interval.  However, I’m unable to reproduce it under normal condition with statusStatusTimer at 2500ms.  I believe it will be very difficult to encounter under normal usage.  

Test report for test#4 can be download here:
https://drive.google.com/file/d/1CJuaxPktqSTsD7N0J8CRyJ8JSVkMuBa_/view?usp=sharing

**Comment by Ryzee119** [comment]
Once again, thank for you for the thorough testing.
I have push some more changes to hopefully fix that LED issue. These wireless controllers are very picky 👍 

The simply the act of logging is very slow may be causing more missed events than without logging.  But I think at this point there's not too much else we can do without alot of rewrite. We're at an acceptable level now I hope.

**Comment by phantom-8** [comment]
I've retest the LED issue, it's now fixed.

For test case #4, I first tested without logging.  When problem occurred, I turned on logging (by adding the logging code 'cos it wasn't in the latest source version).  Then, I re-tested again and collected the log, so I can see what exactly was going on.  Therefore, the problem was not resulted from the addition of logging code.

Anyway, I fully agree with you.  The current version is very good and I wouldn't worry about the possibility of checkControllerPresence() causing missing button events.  In fact, I wasn't able to reproduce the problem under normal 2500ms timing.

Again, thank you very much for your excellent work and the help to solving these issues.

**Comment by Ryzee119** [comment]
No problem ☺ checkControllerPresence() may even be optional. Could just remove it too

**Comment by phantom-8** [comment]
If it can be removed, it will be perfect!  Anyway, I'm extremely happy with the current version.  It's your call whether you still want to try removing it.

I've just received the 3.3V/16MHz Pro Micro.  I'll test it later and see if it's stable.  I'll also re-test the 3.3V/8MHz version and see if it's fast enough to work reliably.  I'll report back later on.

**Comment by Ryzee119** [comment]
I will make it so that checkControllerPresence is only called if the controller has been idle for some time. No point checking its present if the user is mashing keys.
This should eliminate this being an issue entirely. Keep an eye out for an update in the next day or so

**Comment by phantom-8** [comment]
Will re-test once your updated version is out.

Following are the test results of different Pro Micro versions.

3.3V/16MHz Pro Micro:
- Very stable!  Works exactly like the 5V/16MHz.  Unlike the 5V/16Mz version, the benefit is you don't need additional logic level convertors and it's very easy to build.

3.3V/8MHz Pro Micro:
- I recompiled with F_CPU clock set to 8MHz and didn't touch any other source files.  That is, the SPI bus will run at 2MHz comparing to 4MHz for other Pro Micro versions.  Even tough the chip is running at half of speed, it's totally capable of handling all the required task.
- I tested it under heavy load with calling chatPadKeepAlive() every 10ms and no missed button events.

**Comment by jimnarey** [comment]
@phantom-8 With your 8Mhz/3.3V version how are you getting 5V to the USB Shield VBUS? Can you just solder a wire from +5V on the Micro's USB connector to the positive pin on the shield's USB connector? 

I put together a version using a 16Mhz/5V Pro Micro and mini shield but in addition to a level shifter (they all seem to be 4 channel, too small, or eight channel, more than needed) it needed a 3.3V regulator. So it was a lot of additional hardware and complexity for relatively little gain.

My next idea is to try compiling a version without the host shield lib/code for the mega16u2 to try and make this work with an Uno (with the host shield code then on the 328p). With hoodloader2 and various other bits, all of the required elements are out there but I'm not quite sure how to approach changing the board settings in Lufa. I'm getting errors compiling twi (presumably because the mega16u2 doesn't have it). 

If either of you have any pointers that would be great! This isn't an issue with ogx360 so didn't want to post it as one...

Looking forward to trying the 32u4 branch!

**Comment by phantom-8** [comment]
@jimnarey, on the Pro Micro, the RAW pin is USB power.  It goes through a diode behind the USB power pin, there would be a 0.5V drop and voltage is actually 4.5V.  It’s enough to power most USB peripherals.  Testing aside, I would not build a 5V/16MH Pro Micro version, as it needs level shifter and it will be very complicate to build.

Here are the photos of my build.  The same connections apply to both 3.3V/16MHz & 3.3V/8MHz.  The only difference is you need to change the F_CPU macro and recompile the firmware for 3.3V/8MHz.

Connection Details:
- You’ll need to cut a trace on the Mini USB Shield.  By default, the USB power on the mini shield is connected to 3.3V.  By cutting the trace, you can supply 5V to the USB Power by connecting 5V (from Pro Micro RAW pin) to the shield’s USB Power Pin (red wire on my photo).  WARNING: DO NOT SUPPLY 5V POWER TO THE RED WIRE WITHOUT CUTTING THE TRACE.  DOING SO WILL BURN OUT BOTH YOUR SHIELD & PRO MICRO.  Use a meter to confirm the trace has been severed successfully.
- Solder the pin headers & 2 wires (red & black) on the mini shield as shown on the photo.
- Solder the other end of the red wire to the bottom of RAW pin hole on your Pro Micro.
- Solder the other end of the black wire to the bottom of D8 (labelled as 8) on Pro Micro.
- Stack the Pro Micro on top of the shield and finish soldering the end of the pin headers.
- You’ll also need to change the macro USB_HOST_RESET_PIN to 8 in settings.h file.  Line 12 should be like the following after modification.  See below explanation for reason of doing this change.
  #define USB_HOST_RESET_PIN 8
- For 3.3V/8MHz version, you'll need to change the F_CPU macro in file CMakeLists.txt.  Line 4 should now read as below.  No need to make this change if you're using 3.3V/16MHz Pro Micro.
  set(F_CPU 8000000UL)

Notes:
- The connections above is a bit different than the original ogx360 schematics.  I purposely reroute the Shield’s RESET pin to D8, not D9 on the ogx360 schematics.  I connect D9 to INT pin of the MAX chip.  As Ryzee119 has mentioned in the UHS3.0 enhancement thread, he has plans to support UHS3.0 and it requires the use of INT pin on the MAX.  This is why I pre-wire D9 to INT.  It will not be used at the moment.  In the future, if UHS3.0 is supported, we don’t need to re-solder.  However, if you prefer to follow the current ogx360 schematics, you can simply not to install a pin header on the shield’s INT pin and solder the other end of the black wire to Pro Micro’s D9.  Then, you don’t need to modify the USB_HOST_RESET_PIN mapping on the source file.

Hope this help.

![min_shield_trace](https://user-images.githubusercontent.com/8468733/101268929-e7051900-37a3-11eb-8e4d-4ceee4ff09e1.jpg)
![IMG_20201206_082705](https://user-images.githubusercontent.com/8468733/101268952-229fe300-37a4-11eb-8620-51291f87b9b7.jpg)
![IMG_20201206_082759](https://user-images.githubusercontent.com/8468733/101268997-7ad6e500-37a4-11eb-9fc6-1770bc11d63c.jpg)
![IMG_20201206_090607](https://user-images.githubusercontent.com/8468733/101268957-33505900-37a4-11eb-934a-19c778ddc6b5.jpg)
![IMG_20201206_090619](https://user-images.githubusercontent.com/8468733/101268961-38ada380-37a4-11eb-8b3d-fe5d95a5be95.jpg)

**Comment by phantom-8** [comment]
@jimnarey, for your 2nd question regarding LUFA, I'm afraid I'm unable to answer.  This whole USB / LUFA is new to me.  I've just stared to learn the USB protocol while doing the handshake with the 360 wired controller.  Maybe Ryzee119 can help you in this area.

**Comment by phantom-8** [comment]
@Ryzee119, I noticed there is a new commit on checking for presence when idle only.  In file XBOXRECV.cpp, at line 297, you reset the idleTimer by checking bufferSize of the inTransfer.  For the controller with chatpad, chatPadKeepAliveN() will be called regularly and it will result in several messages response from the controller.  This will in turn reset the idleTimer.  In other words, the idle timer will be reset each time chatPadKeepAliveN() is called.  Is this intentional?  If my calculation is correct, chatPadKeepAlive1() is called every 7500ms and chatPatKeepAlive2() is called 2500ms right after chatPadKeepAlive2().  Or is it better to reset idleTimer when buttons changes are detected?  I’m just curious.  If you think the current version is fine, I will proceed with testing.  Pls advise.

**Comment by Ryzee119** [comment]
Thats still a work in progress. Some reason its resetting the boards. 😒

**Comment by phantom-8** [comment]
I see.  I'll wait for your confirmation before testing.  If you need help in testing board reset, let me know.

**Comment by Ryzee119** [comment]
Alright, so it now only sends presense checks if the controller isnt already connected, or the user hasnt pressed a button in 2.5 seconds.
I had some issue with stack corruption. I think we're running out of RAM.

Sorry to be annoying, but I migrated the entire build system to PlatformIO. This uses the latest and complete ArduinoCore and seems to have resolved the stack corruption issues. To change the FCPU speed you now just modify `platformio.ini` respective variable.

`xboxrecv` branch has all been updated.

**Comment by jimnarey** [comment]
@phantom-8 Thank you so much for the detailed instructions!

I've had a go at building one but have clearly gone wrong somewhere. It didn't pick up the controller, when connected (no LED/sign of life on the controller) and subsequently wouldn't be recognised in dmesg as a an xinput device. I'm going to have another go tomorrow. I don't doubt the instructions are correct.

@Ryzee119 - looking forward to trying the PlatformIO build!

**Comment by phantom-8** [comment]
@Ryzee119 , I'll need to install PlatformIO first.  Will test the new version in a moment.  Looks like there are some big changes.

@jimnarey, which version did you build?  3.3V/16MHz or 3.3V/8Hz.  Can you post some pics of your build?

**Comment by phantom-8** [comment]
@Ryzee119, idleTimer is not working.  In file XBOXRECV.cpp line 448
```
448                memset(readBuf, 0x00, 32);
```
This will clear the readBuf.  The check on line 302 will always fail and idleTimer will never to reset.
```
302                if (readBuf[1] & 0x01)
303                     idleTimer[i] = millis();

I'm still testing other functions.  Will post more later.

**Comment by Ryzee119** [comment]
That memset should only occur after the report is parsed.
It should be repopulated before that idle check. 🤔

**Comment by phantom-8** [comment]
The memset is at the end of readReport().  When readReport returns, the readBuf will be cleared.  The check
```
if (readBuf[1] & 0101)
```
This check will always fail, 'cos readBuf was cleared by memset (all zeroes).  So, idleTimer will never be set.

**Comment by Ryzee119** [comment]
readbuff should be repopulated before that check on
https://github.com/Ryzee119/ogx360/blob/d1b0d782838d27f11d4f10e7b2c87684c1001eac/Firmware/src/lib/UHS/XBOXRECV.cpp#L297

**Comment by phantom-8** [comment]
No problem :)

**Comment by phantom-8** [comment]
I guess it's easier to do the check inside readReport().  In that case, you don't have to take care of resetting idleTimer at 2 places, around line 302 & line 593.

**Comment by phantom-8** [comment]
Time for another test report.  The idleTimer is now working as expected.  I'm now unable to reproduce controller button lost event under any circumstances.  Excellent work!

Previously, I reported using Guide button to connect sometimes resulted in LED kept flashing.  You had fixed it by regularly calling setLedRaw().  It would take a bit longer if it couldn't set the LED the first time.  Now, this new version doesn't has this problem.  I noticed that the LED controller number always synced the 1st time.  I didn't know what you did, but it works much better than before.  

I've performed additional testing with Steel Battalion.  For some reason, chatpad button events were missing once a while.  Also, rumble sometimes didn't stop.  Not sure if these problems were there before 'cos I didn't tested SB extensively before this firmware update.

**Comment by phantom-8** [comment]
Okay, to reproduce the chatpad event lost.  Wait until the chatpad LED goes off, then immediately press a chatpad button.  There is a high chance the chatpad button won't be detected.  I can simulate the problem without SB running and with additional logging.  From the log I obtained, there are no chatpad button message reported back when it's lost.  Very strange!

**Comment by phantom-8** [comment]
I think I found the problem.  Currently, the chatPadKeepAlive() are being called every 7500ms.  That chatpad hardware actually timed out and won't send any button press.  I now tried to call chatPadKeepAlive() every 2500ms, and no more lost chatpad button event.

**Comment by Ryzee119** [comment]
Excellent 😊 so just that rumble issue?

The steel Battalion rumble hooks the led brightness feedback to drive the rumble motors because rumble isnt actually supported by SB.

I hook things like chaff and eject button leds. 

Does it seem like the rumble issues could be a side effect of this or it just genuinely has issues? Ill investigate when i can

**Comment by phantom-8** [comment]
The rumble issue is still there, but it's no big deal.  The current firmware is really incredible and be able to play SB LOC with my daughter is godsend.  Thanks again for the wonderful piece of work.

**Comment by jimnarey** [comment]
@phantom-8 Thank you.

Here are the pictures of the build. The ribbon cable is for an OLED (which is working on the Leonardo). I tried a hex without the OLED enabled to see if that was the problem but am going to build another today without the ribbon, just in case. The OLED itself seemed to be OK.

I'm also wondering whether I might have damaged it plugging a PS4 controller in (which has a max 800mA draw) - again working fine on the Leonardo. So I'll go back to first principles on it and make sure I try a more basic hex as well as not modify the build in any way.

![WhatsApp Image 2020-12-08 at 13 31 26](https://user-images.githubusercontent.com/6508352/101490115-d26a8180-3959-11eb-88dc-f2b3b2a0ca01.jpeg)

![WhatsApp Image 2020-12-08 at 13 31 27](https://user-images.githubusercontent.com/6508352/101490131-d9918f80-3959-11eb-9cf1-13f804c16e40.jpeg)

**Comment by phantom-8** [comment]
Your build looks fine.  Although I couldn’t see the 5V & RESET wiring, I assumed they’re okay.  Did you try to measure the USB Connector Power Pin on the Shield to confirm it’s outputting 4.5V?

Did you change the USB_HOST_RESET_PIN macro to 8?  It won’t work with original setting to 9.

Which Pro Micro are you using?  3.3V/16MHz or 3.3V/8MHz?  Let me know and I’ll provide you a pre-build binary for testing.

**Comment by jimnarey** [comment]
@phantom-8  Hi - thank you.

Yes, I changed the pin to 8 and changed the clock speed. I'm using a 3.3V/*Mhz.

A binary to test would be really helpful, thank you!

**Comment by phantom-8** [comment]
@jimnarey, I believe you meant 3.3V/8MHz.  Here is the 8MHz firmware for MASTER_360WLESS_360W_ONEW.


[ogx360fw.zip](https://github.com/Ryzee119/ogx360/files/5684742/ogx360fw.zip)

**Comment by jimnarey** [comment]
I did! Thank you for this and all your previous help!

**Comment by UnElectronicoMas** [comment]
@Ryzee119 @phantom-8 @jimnarey 

hi guys, i was trying to make the ogx360 for one controller and i bought the arduino pro mini ( atmega 328p 3.3/8mhz) for error instead of arduino pro micro 3.3/8 mhz, my question is if i extract the 3.3 voltage regulator from Arduino pro Mini and i solder it on my arduino pro micro 5v/16Mhz (because i have one but i haven't level shifters), will i need to use level shifters ? or that way can i make my own arduino pro micro 3.3v/16mhz ? i hope you can help me, excellent work by the way

**Comment by hpx1984** [comment]
@phantom-8 
Hi Sir,
 i used ogx360fw.zip above but it can't support xbox360 keyboard, could you use the latest FW from Ryzee119's project for recompile a 8mHz version?
i don't know how the recompile a Hex file
sorry for trouble you，sir

**Comment by hpx1984** [comment]
> @jimnarey, I believe you meant 3.3V/8MHz. Here is the 8MHz firmware for MASTER_360WLESS_360W_ONEW.
> 
> [ogx360fw.zip](https://github.com/Ryzee119/ogx360/files/5684742/ogx360fw.zip)

@phantom-8 
Hi Sir 
could you share the FW 3.3v 16mhz？
